Snowplow/SPEmitter.m

```diff
@@ -172,27 +202,81 @@ - (void) sendEvents {
         for (int i = 0; i < listValues.count; i += _bufferOption) {
             NSMutableArray *eventArray = [[NSMutableArray alloc] init];
             NSMutableArray *indexArray = [[NSMutableArray alloc] init];
-            double stm = [SPUtils getTimestamp];
+            NSInteger stm = [SPUtilities getTimestamp];
+            NSInteger totalByteSize = 0;
             
             for (int j = i; j < (i + _bufferOption) && j < listValues.count; j++) {
+                // Add STM to payload
                 NSMutableDictionary *eventPayload = [[listValues[j] objectForKey:@"eventData"] mutableCopy];
-                [eventPayload setValue:[NSString stringWithFormat:@"%.0f", stm] forKey:kSPSentTimestamp];
-                [eventArray addObject:eventPayload];
-                [indexArray addObject:[listValues[j] objectForKey:@"ID"]];
+                [eventPayload setValue:[NSString stringWithFormat:@"%ld", (long)stm] forKey:kSPSentTimestamp];
+                
+                // Convert to NSData and check the byte size
+                NSData *data = [NSJSONSerialization dataWithJSONObject:eventPayload options:0 error:nil];
+                NSInteger payloadByteSize = [SPUtilities getByteSizeWithString:[[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding]];
+                
+                if (payloadByteSize > _byteLimitPost) {
+                    // Single event exceeds the byte limit so must be sent individually.
+                    NSMutableArray *singleEventArray = [[NSMutableArray alloc] init];
+                    NSMutableArray *singleIndexArray = [[NSMutableArray alloc] init];
+                    
+                    // Reset the STM as it is going individually
+                    [eventPayload setValue:[NSString stringWithFormat:@"%ld", (long)[SPUtilities getTimestamp]] forKey:kSPSentTimestamp];
+                    
+                    // Build and Send the event!
+                    [singleEventArray addObject:eventPayload];
+                    [singleIndexArray addObject:[listValues[j] objectForKey:@"ID"]];
+                    SPSelfDescribingJson *payload = [[SPSelfDescribingJson alloc] initWithSchema:kSPPayloadDataSchema
+                                                                                         andData:singleEventArray];
+                    [self sendEventWithRequest:[self getRequestPostWithData:payload] andIndex:singleIndexArray andResultArray:sendResults andOversize:YES];
+                } else if (totalByteSize + payloadByteSize > _byteLimitPost) {
+                    // Adding this event to the accumulated array would exceed the limit.
```

Comment:
One potential optimization which is out of scope for this release: grouping events to minimize the total number of batches.

For instance, if the byte limit is 10 and we have events of size 6, 6, 4, and 4, then filling creating a new batch whenever the next size would push the current batch over the limit yields 3 batches ({6}, {6}, and {4, 4}). We could get 2 batches ({6, 4} and {6, 4}) using the first fit algorithm described here: https://en.wikipedia.org/wiki/Bin_packing_problem

This probably wouldn't have any impact for users not sending large events.